capstone/src/main/java/com/google/step/servlets/UserSignUpServlet.java

```diff
@@ -0,0 +1,54 @@
+package com.google.step.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that will put the user information in Datastore.
+ */
+@WebServlet("/userSignUp")
+public class UserSignUpServlet extends HttpServlet {
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+      return;
+    }
+
+    String userName = request.getParameter("userName");
+    String id = userService.getCurrentUser().getUserId();
+    String email = userService.getCurrentUser().getEmail();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Check if user email already exists.
+    Query query = new Query("User").setFilter(
+        new Query.FilterPredicate("email", Query.FilterOperator.EQUAL, email));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
```

Comment:
This is not something that we should probably worry about now, but the asSingleEntity() method throws an exception if more than one entity is returned. Because we are not using atomic operations of any sort, we cannot currently guarantee that there won't be a case where two users get to the check at the same time, which could cause two entities with the same "email" field to be placed in the Datastore. This would result in any subsequent sign ups with that email crashing our server, which would not be good.

Based on what I've heard in the past, we would use Transactions to fix this. As I said before, though this probably isn't something we should worry about fixing until later, depending on how simple it is to implement a transactional system.

Reply:
Yeah I think we talked about implementing transactions in the future, this should still work in the meanwhile we get the basics of the app up and running.

Reply:
I think "using transactions to guarantee uniqueness" is a pretty important thing to do, and doesn't take much additional work to get right. Might as well just do it now.

Reply:
Currently not worrying about that because we deleted the servlet and changed the user-restaurant interaction.